File: frontend/src/api/queues.ts
import { api } from "./client";
import type { Queue, QueueCreate } from "../types";

export interface QueueFilters {
  group?: string;
  teacher?: string;
  status?: string;
  date?: string;
}

/** Учебные очереди. Контракт backend — [BE-Q1]..[BE-Q4]. Экраны — [FE-06]. */
export const queuesApi = {
  list: (filters?: QueueFilters) =>
    api.get<Queue[]>("/queues", { params: filters }).then((r) => r.data),

  get: (id: string | number) =>
    api.get<Queue>(`/queues/${id}`).then((r) => r.data),

  create: (body: QueueCreate) =>
    api.post<Queue>("/queues", toBackendQueueCreate(body)).then((r) => r.data),

  join: (id: string | number) =>
    api.post<Queue>(`/queues/${id}/join`).then((r) => r.data),

  leave: (id: string | number) =>
    api.post<Queue>(`/queues/${id}/leave`).then((r) => r.data),

  reorder: (id: string | number, memberIds: (string | number)[]) =>
    api
      .patch<Queue>(`/queues/${id}/members/reorder`, { order: memberIds })
      .then((r) => r.data),

  removeMember: (id: string | number, memberId: string | number) =>
    api.delete<Queue>(`/queues/${id}/members/${memberId}`).then((r) => r.data),

  /** Отметить участника сдавшим, опционально с оценкой. */
  completeMember: (id: string | number, memberId: string | number, grade: number | null) =>
    api
      .patch<Queue>(`/queues/${id}/members/${memberId}/complete`, { grade })
      .then((r) => r.data),

  close: (id: string | number) =>
    api.patch<Queue>(`/queues/${id}/close`).then((r) => r.data),
};

function toBackendQueueCreate(body: QueueCreate) {
  const { date, time_start } = parseWhen(body.when);
  return {
    title: body.title,
    discipline: body.discipline,
    qtype: body.qtype,
    group: body.group,
    room: body.room,
    date,
    time_start,
    max: body.max,
    comment: body.comment,
  };
}

function parseWhen(value: string): { date: string; time_start: string } {
  const match = value.match(/(\d{2})\.(\d{2})(?:\.(\d{4}))?\s*[·,]?\s*(\d{1,2}):(\d{2})/);
  if (!match) {
    const today = new Date().toISOString().slice(0, 10);
    return { date: today, time_start: "09:00:00" };
  }
  const [, dd, mm, yyyy, hh, min] = match;
  return {
    date: `${yyyy ?? new Date().getFullYear()}-${mm}-${dd}`,
    time_start: `${hh.padStart(2, "0")}:${min}:00`,
  };
}

File: backend/app/schemas/queues.py
"""Pydantic-схемы для queues-эндпоинтов.

Контракт совпадает с TypeScript-типами Queue, QueueMember, QueueCreate на фронтенде.
"""
from __future__ import annotations

import uuid
from datetime import date, time

from pydantic import BaseModel, Field

from app.models.queue import QueueStatus, QueueType


class QueueMemberOut(BaseModel):
    id: int
    userId: uuid.UUID
    name: str = ""
    seq: int
    position: int
    passed: bool = False
    grade: int | None = None

    model_config = {"from_attributes": True}


class QueueOut(BaseModel):
    id: int
    title: str
    discipline: str
    qtype: str
    teacher: str = ""
    room: str
    group: str
    when: str          # ISO datetime строка для фронтенда
    max: int
    status: str
    comment: str | None = None
    students: list[QueueMemberOut] = []

    model_config = {"from_attributes": True}


class QueueCreate(BaseModel):
    title: str
    discipline: str
    qtype: QueueType
    group: str
    room: str
    date: date
    time_start: time
    time_end: time | None = None
    max: int = Field(default=30, ge=1, le=200)
    comment: str | None = None


class ReorderRequest(BaseModel):
    order: list[int]  # список member_id в новом порядке
